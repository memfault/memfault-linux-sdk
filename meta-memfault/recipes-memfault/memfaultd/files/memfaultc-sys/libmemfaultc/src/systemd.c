//! @file
//!
//! Copyright (c) Memfault, Inc.
//! See License.txt for details
//!
//! @brief
//! memfaultd systemd helper

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <systemd/sd-bus.h>

static const char *const systemd_service = "org.freedesktop.systemd1";

/**
 * @param state output variable for current state (to be freed by caller).
 */
static bool prv_systemd_get_service_state(sd_bus *bus, const char *service_name, char **state) {
  static const char *const unit_interface = "org.freedesktop.systemd1.Unit";

  char *unit_path = NULL;
  bool result = true;

  if (sd_bus_path_encode("/org/freedesktop/systemd1/unit", service_name, &unit_path) < 0) {
    fprintf(stderr, "memfaultd:: Failed to generate SystemD unit path\n");

    return false;
  }

  sd_bus_error error = SD_BUS_ERROR_NULL;
  if (sd_bus_get_property_string(bus, systemd_service, unit_path, unit_interface, "ActiveState",
                                 &error, state) < 0) {
    fprintf(stderr, "memfaultd:: Failed to get state of %s: %s\n", service_name, error.name);
    sd_bus_error_free(&error);
    result = false;
  }

  if (unit_path) {
    free(unit_path);
  }
  return result;
}

/*
 * List of SystemD commands: https://www.freedesktop.org/wiki/Software/systemd/dbus/
 */

static bool prv_systemd_restart_service(sd_bus *bus, const char *service_name) {
  const char *manager_path = "/org/freedesktop/systemd1";
  const char *manager_interface = "org.freedesktop.systemd1.Manager";

  sd_bus_error error = SD_BUS_ERROR_NULL;
  sd_bus_message *msg = NULL;

  if (sd_bus_call_method(bus, systemd_service, manager_path, manager_interface, "RestartUnit",
                         &error, &msg, "ss", service_name, "replace") < 0) {
    fprintf(stderr, "memfaultd:: Failed to restart %s: %s\n", service_name, error.name);
    sd_bus_error_free(&error);
    return false;
  }

  sd_bus_message_unref(msg);

  return true;
}

static bool prv_systemd_kill_service(sd_bus *bus, const char *service_name, int signal) {
  const char *manager_path = "/org/freedesktop/systemd1";
  const char *manager_interface = "org.freedesktop.systemd1.Manager";

  sd_bus_error error = SD_BUS_ERROR_NULL;
  sd_bus_message *msg = NULL;

  /*
    Refer to SystemD Bus documentation for arguments to the call:
      KillUnit(in  s name,
               in  s who,       // "all": is the default (like systemctl kill service)
               in  i signal);
  */
  if (sd_bus_call_method(bus, systemd_service, manager_path, manager_interface, "KillUnit", &error,
                         &msg, "ssi", service_name, "all", signal) < 0) {
    fprintf(stderr, "memfaultd:: Failed to kill %s: %s\n", service_name, error.name);
    sd_bus_error_free(&error);
    return false;
  }

  sd_bus_message_unref(msg);

  return true;
}

/**
 * @brief Restart service using systemd dbus API if already running
 *
 * @param src_module Friendly name of caller module
 * @param service_name Service name to restart, e.g. collectd.service
 * @return true Successfully restarted requested service
 * @return false Failed to restart service
 */
bool memfaultd_restart_systemd_service_if_running(const char *service_name) {
  bool result = true;

  // Initialize connection to SystemD
  sd_bus *bus;
  if (sd_bus_default_system(&bus) < 0) {
    fprintf(stderr, "memfaultd:: Failed to find systemd system bus\n");
    return false;
  }

  // Check if service is active before restarting it
  char *state = NULL;
  if (!prv_systemd_get_service_state(bus, service_name, &state)) {
    result = false;
    goto cleanup;
  }
  if (strcmp("active", state) != 0 && strcmp("activating", state) != 0) {
    fprintf(stderr, "memfaultd:: %s is not active (%s). Not starting.\n", service_name, state);
    result = true;
    goto cleanup;
  }

  // Restart the service
  if (!prv_systemd_restart_service(bus, service_name)) {
    result = false;
    goto cleanup;
  }

cleanup:
  if (state != NULL) {
    free(state);
  }
  sd_bus_unref(bus);
  return result;
}

/**
 * @brief Send a signal to a service if it's running.
 *
 * @param src_module Friendly name of caller module
 * @param service_name Service name to "kill", e.g. collectd.service
 * @param signal Signal to send, e.g. SIGUSR1
 * @return true Successfully sent signal to requested service
 * @return false Failed to restart service
 */
bool memfaultd_kill_systemd_service(const char *service_name, int signal) {
  bool result = true;

  // Initialize connection to SystemD
  sd_bus *bus;
  if (sd_bus_default_system(&bus) < 0) {
    fprintf(stderr, "memfaultd:: Failed to find systemd system bus\n");
    return false;
  }

  // Send signal to service
  if (!prv_systemd_kill_service(bus, service_name, signal)) {
    result = false;
    goto cleanup;
  }

cleanup:
  sd_bus_unref(bus);
  return result;
}

/**
 * @brief Checks if the current systemd state matches the requested state
 *
 * @return NULL Failed to get current systemd state
 *
 * @note Caller must free returned string
 */
char *memfaultd_get_systemd_bus_state() {
  sd_bus *bus = NULL;
  sd_bus_error error = SD_BUS_ERROR_NULL;
  char *cur_state = NULL;

  const char *service = "org.freedesktop.systemd1";
  const char *path = "/org/freedesktop/systemd1";
  const char *interface = "org.freedesktop.systemd1.Manager";
  const char *system_state = "SystemState";

  const int bus_result = sd_bus_default_system(&bus);
  if (bus_result < 0) {
    fprintf(stderr, "reboot:: Failed to find systemd system bus: %s\n", strerror(-bus_result));
    goto cleanup;
  }

  if (sd_bus_get_property_string(bus, service, path, interface, system_state, &error, &cur_state) <
      0) {
    fprintf(stderr, "reboot:: Failed to get system state: %s\n", error.name);
    goto cleanup;
  }

cleanup:
  sd_bus_error_free(&error);
  sd_bus_unref(bus);
  return cur_state;
}
