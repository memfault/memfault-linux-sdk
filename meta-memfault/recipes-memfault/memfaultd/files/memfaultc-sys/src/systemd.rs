//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::ffi::c_int;

use libc::{c_char, size_t};

#[allow(non_camel_case_types)]
pub enum sd_journal {}

extern "C" {
    pub fn memfaultd_restart_systemd_service_if_running(service_name: *const c_char) -> bool;
    pub fn memfaultd_get_systemd_bus_state() -> *const c_char;
    pub fn sd_journal_open(ret: *mut *mut sd_journal, flags: c_int) -> c_int;
    pub fn sd_journal_seek_tail(j: *mut sd_journal) -> c_int;
    pub fn sd_journal_previous(j: *mut sd_journal) -> c_int;
    pub fn sd_journal_next(j: *mut sd_journal) -> c_int;
    pub fn sd_journal_get_data(
        j: *mut sd_journal,
        field: *const c_char,
        data: *mut *mut u8,
        l: *mut size_t,
    ) -> c_int;
    pub fn sd_journal_enumerate_data(
        j: *mut sd_journal,
        data: *mut *mut u8,
        l: *mut size_t,
    ) -> c_int;
    pub fn sd_journal_get_fd(j: *mut sd_journal) -> c_int;
    pub fn sd_journal_process(j: *mut sd_journal) -> c_int;
    pub fn sd_journal_get_realtime_usec(j: *mut sd_journal, ret: *mut u64) -> c_int;
    pub fn sd_journal_get_cursor(j: *mut sd_journal, cursor: *mut *const c_char) -> c_int;
    pub fn sd_journal_seek_cursor(j: *mut sd_journal, cursor: *const c_char) -> c_int;
    pub fn sd_journal_add_match(j: *mut sd_journal, data: *const c_char, size: size_t) -> c_int;
}
