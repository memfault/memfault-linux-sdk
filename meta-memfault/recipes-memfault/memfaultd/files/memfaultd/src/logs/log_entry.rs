//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::collections::HashMap;

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};

use crate::util::serialization::datetime_to_rfc3339;

const MESSAGE_KEY: &str = "MESSAGE";
const PID_KEY: &str = "_PID";
const SYSTEMD_UNIT_KEY: &str = "_SYSTEMD_UNIT";
const PRIORITY_KEY: &str = "PRIORITY";
const ORIGINAL_PRIORITY_KEY: &str = "ORIGINAL_PRIORITY";

#[derive(Debug, Serialize, Deserialize, Clone)]
#[serde(untagged)]
pub enum LogValue {
    String(String),
    Float(f64),
}

#[derive(Debug, Serialize, Deserialize, Clone)]
/// Represents the structured format of a log entry
///
/// Note that we will not serialize the fields that are `None` to save space.
pub struct LogData {
    #[serde(rename = "MESSAGE")]
    pub message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "_PID")]
    pub pid: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "_SYSTEMD_UNIT")]
    pub systemd_unit: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "PRIORITY")]
    pub priority: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(rename = "ORIGINAL_PRIORITY")]
    pub original_priority: Option<String>,
    #[serde(flatten)]
    pub extra_fields: HashMap<String, LogValue>,
}

impl LogData {
    /// Returns the value of the field with the given key.
    pub fn get_field(&self, key: &str) -> Option<String> {
        match key {
            MESSAGE_KEY => Some(self.message.clone()),
            PID_KEY => self.pid.clone(),
            SYSTEMD_UNIT_KEY => self.systemd_unit.clone(),
            PRIORITY_KEY => self.priority.clone(),
            ORIGINAL_PRIORITY_KEY => self.original_priority.clone(),
            _ => self.extra_fields.get(key).and_then(|v| match v {
                LogValue::String(s) => Some(s.clone()),
                LogValue::Float(_) => None,
            }),
        }
    }
}

/// Represents a structured log that could come from a variety of sources.
#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct LogEntry {
    #[serde(with = "datetime_to_rfc3339")]
    pub ts: DateTime<Utc>,
    pub data: LogData,
}

impl LogEntry {
    /// Filter log fields to only include defaults and those specified in `extra_fields`.
    ///
    /// This function modifies the log entry data to remove any extra fields that are not
    /// specified by the user.
    pub fn filter_extra_fields(&mut self, extra_fields: &[String]) {
        self.data
            .extra_fields
            .retain(|k, _| extra_fields.contains(k));
    }
}

#[cfg(test)]
impl LogEntry {
    pub fn new_with_message(message: &str) -> Self {
        LogEntry {
            ts: Utc::now(),
            data: LogData {
                message: message.to_string(),
                pid: None,
                systemd_unit: None,
                priority: None,
                original_priority: None,
                extra_fields: HashMap::new(),
            },
        }
    }

    pub fn new_with_message_and_ts(message: &str, ts: DateTime<Utc>) -> Self {
        LogEntry {
            ts,
            data: LogData {
                message: message.to_string(),
                pid: None,
                systemd_unit: None,
                priority: None,
                original_priority: None,
                extra_fields: HashMap::new(),
            },
        }
    }
}

#[cfg(test)]
mod tests {
    use chrono::{DateTime, TimeZone, Utc};
    use insta::{assert_json_snapshot, with_settings};
    use rstest::{fixture, rstest};

    use super::*;

    #[rstest]
    #[case("only_message", r#"{"MESSAGE":"TEST" }"#, "")]
    #[case("extra_key", r#"{"MESSAGE":"TEST", "SOME_EXTRA_KEY":"XX" }"#, "")]
    #[case(
        "multi_key_match",
        r#"{"MESSAGE":"TEST", "SOME_EXTRA_KEY":"XX", "_PID": "44", "_SYSTEMD_UNIT": "some.service", "PRIORITY": "6" }"#,
         ""
    )]
    #[case(
        "extra_attribute_filter",
        r#"{"MESSAGE":"TEST", "SOME_EXTRA_KEY":"XX" }"#,
        "SOME_EXTRA_KEY"
    )]
    fn test_filtering(
        time: DateTime<Utc>,
        #[case] test_name: String,
        #[case] input: String,
        #[case] extras: String,
    ) {
        let mut entry = LogEntry {
            ts: time,
            data: serde_json::from_str(&input).unwrap(),
        };

        let extra_attributes = extras.split(',').map(String::from).collect::<Vec<_>>();
        entry.filter_extra_fields(&extra_attributes);

        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(test_name, entry);
        });
    }

    #[fixture]
    fn time() -> DateTime<Utc> {
        Utc.timestamp_millis_opt(1334250000000).unwrap()
    }
}
