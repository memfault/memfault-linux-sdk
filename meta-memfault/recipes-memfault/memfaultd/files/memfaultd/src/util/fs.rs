//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use std::{
    fs::{self},
    path::{Path, PathBuf},
};

use eyre::{Context, Result};
use log::warn;

/// Takes a directory and returns a vector of all files in that directory, sorted
/// by creation date:
#[allow(dead_code)] // Required to build without warnings and --no-default-features
pub fn get_files_sorted_by_mtime(dir: &Path) -> Result<Vec<PathBuf>> {
    let read_dir = std::fs::read_dir(dir)?;
    let mut entries = read_dir
        .filter_map(|e| match e {
            Ok(e) => Some(e),
            Err(e) => {
                warn!("Error reading directory entry: {:#}", e);
                None
            }
        })
        .filter(|entry| entry.path().is_file())
        .collect::<Vec<_>>();
    // Order by oldest first:
    entries.sort_by_key(|entry| {
        entry
            .metadata()
            .map(|m| m.modified().unwrap_or(std::time::SystemTime::UNIX_EPOCH))
            .unwrap_or(std::time::SystemTime::UNIX_EPOCH)
    });
    Ok(entries.into_iter().map(|m| m.path()).collect())
}

/// Move a file. Try fs::rename first which is most efficient but only works if source
/// and destination are on the same filesystem.
/// Use Copy/Delete strategy if rename failed.
pub fn move_file(source: &PathBuf, target: &PathBuf) -> Result<()> {
    if fs::rename(source, target).is_err() {
        fs::copy(source, target).wrap_err_with(|| {
            format!(
                "Error moving file {} to {}",
                source.display(),
                target.display()
            )
        })?;
        fs::remove_file(source)?;
    }
    Ok(())
}

/// Copy a file.
///
/// If the source and target are the same, do nothing.
pub fn copy_file(source: &PathBuf, target: &PathBuf) -> Result<()> {
    if source == target {
        return Ok(());
    }

    fs::copy(source, target).wrap_err_with(|| {
        format!(
            "Error copying file {} to {}",
            source.display(),
            target.display()
        )
    })?;
    Ok(())
}
