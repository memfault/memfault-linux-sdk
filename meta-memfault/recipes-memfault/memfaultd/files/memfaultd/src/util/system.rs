//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use eyre::{eyre, Result};
use lazy_static::lazy_static;
use libc::{clockid_t, timespec, CLOCK_MONOTONIC};

#[cfg(target_os = "linux")]
use libc::{sysconf, _SC_CLK_TCK, _SC_PAGE_SIZE};

#[cfg(target_os = "linux")]
use std::fs::read_to_string;

use uuid::Uuid;

lazy_static! {
    static ref OS_RELEASE: Option<String> = read_osrelease().ok();
    static ref OS_TYPE: Option<String> = read_ostype().ok();
}

#[cfg(target_os = "linux")]
pub fn read_system_boot_id() -> Result<Uuid> {
    use eyre::Context;
    use std::str::FromStr;

    const BOOT_ID_PATH: &str = "/proc/sys/kernel/random/boot_id";
    let boot_id = read_to_string(BOOT_ID_PATH);

    match boot_id {
        Ok(boot_id_str) => Uuid::from_str(boot_id_str.trim()).wrap_err("Invalid boot id"),
        Err(_) => Err(eyre!("Unable to read boot id from system.")),
    }
}

#[cfg(target_os = "linux")]
fn read_osrelease() -> Result<String> {
    let mut os_release = read_to_string("/proc/sys/kernel/osrelease")
        .map_err(|e| eyre!("Unable to read osrelease: {}", e))?;
    os_release.truncate(os_release.trim_end().len());
    Ok(os_release)
}

#[cfg(target_os = "linux")]
fn read_ostype() -> Result<String> {
    let mut os_type = read_to_string("/proc/sys/kernel/ostype")
        .map_err(|e| eyre!("Unable to read ostype: {}", e))?;
    os_type.truncate(os_type.trim_end().len());
    Ok(os_type)
}

/// Returns the value stored in /proc/sys/kernel/osrelease
/// The first time it is called we read and cache
/// the value from procfs. On subsequent calls the cached value
/// is returned. This is safe to do since this is a value set
/// at kernel build-time and cannot be altered at runtime.
pub fn get_osrelease() -> Option<String> {
    OS_RELEASE.clone()
}

/// Returns the value stored in /proc/sys/kernel/ostype
/// The first time it is called we read and cache
/// the value from procfs. On subsequent calls the cached value
/// is returned. This is safe to do since this is a value set
/// at kernel build-time and cannot be altered at runtime.
pub fn get_ostype() -> Option<String> {
    OS_TYPE.clone()
}

#[cfg(target_os = "linux")]
pub fn clock_ticks_per_second() -> u64 {
    unsafe { sysconf(_SC_CLK_TCK) as u64 }
}

#[cfg(target_os = "linux")]
pub fn bytes_per_page() -> u64 {
    unsafe { sysconf(_SC_PAGE_SIZE) as u64 }
}

/// Calls clock_gettime
/// Most interesting to us are:
/// CLOCK_MONOTONIC: "clock that increments monotonically, tracking the time
/// since an arbitrary point, and will continue to increment while the system is
/// asleep."
/// CLOCK_BOOTTIME  A  nonsettable system-wide clock that is identical to
/// CLOCK_MONOTONIC, except that it also includes any time that the system is
/// suspended.  This allows applications to get a suspend-aware monotonic clock
/// without having to deal with the complications of CLOCK_REALTIME, which may
/// have discontinuities if the time is changed using settimeofday(2) or
/// similar.
pub enum Clock {
    Monotonic,
    Boottime,
}
pub fn get_system_clock(clock: Clock) -> Result<std::time::Duration> {
    // Linux only so we define it here.
    const CLOCK_BOOTTIME: clockid_t = 7;

    let mut t = timespec {
        tv_sec: 0,
        tv_nsec: 0,
    };
    if unsafe {
        libc::clock_gettime(
            match clock {
                Clock::Monotonic => CLOCK_MONOTONIC,
                Clock::Boottime if cfg!(target_os = "linux") => CLOCK_BOOTTIME,
                // Falls back to monotonic if not linux
                Clock::Boottime => CLOCK_MONOTONIC,
            },
            &mut t,
        )
    } != 0
    {
        Err(eyre!("Error getting system clock."))
    } else {
        Ok(std::time::Duration::new(t.tv_sec as u64, t.tv_nsec as u32))
    }
}

/// Provide some mock implementations for non-Linux systems. Designed for development. Not actual use.

#[cfg(not(target_os = "linux"))]
pub fn read_system_boot_id() -> Result<Uuid> {
    use once_cell::sync::Lazy;
    static MOCK_BOOT_ID: Lazy<Uuid> = Lazy::new(Uuid::new_v4);
    Ok(*MOCK_BOOT_ID)
}

#[cfg(not(target_os = "linux"))]
pub fn read_osrelease() -> Result<String> {
    Ok("non-Linux-system".to_string())
}

#[cfg(not(target_os = "linux"))]
pub fn read_ostype() -> Result<String> {
    Ok("non-Linux-system".to_string())
}

#[cfg(not(target_os = "linux"))]
pub fn clock_ticks_per_second() -> u64 {
    10_000
}

#[cfg(not(target_os = "linux"))]
pub fn bytes_per_page() -> u64 {
    4096
}
