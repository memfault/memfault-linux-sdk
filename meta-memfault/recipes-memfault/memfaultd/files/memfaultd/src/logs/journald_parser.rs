//
// Copyright (c) Memfault, Inc.
// See License.txt for details
use chrono::{DateTime, NaiveDateTime, Utc};
use eyre::{eyre, Error, Result};
use libc::free;
use nix::poll::{poll, PollFd};
use serde::Serialize;
use std::ffi::CString;
use std::fs::read_to_string;
use std::{collections::HashMap, path::PathBuf};
use std::{ffi::c_char, mem::MaybeUninit};

use log::{debug, warn};
use memfaultc_sys::systemd::{
    sd_journal, sd_journal_add_match, sd_journal_enumerate_data, sd_journal_get_cursor,
    sd_journal_get_fd, sd_journal_get_realtime_usec, sd_journal_next, sd_journal_open,
    sd_journal_process, sd_journal_seek_cursor,
};

use super::log_entry::{LogData, LogEntry, LogValue};
use crate::util::system::read_system_boot_id;
/// A trait for reading journal entries from the systemd journal.
///
/// This trait is used to abstract the raw journal entry reading logic from the rest of the codebase.
/// This allows for easier testing and mocking of the journal reading logic.
#[cfg_attr(test, mockall::automock)]
pub trait JournalRaw {
    /// Check if the next journal entry is available.
    ///
    /// Returns `Ok(true)` if the next journal entry is available, `Ok(false)` if there are no more entries,
    /// and an error if there was an issue reading the journal.
    fn next_entry_available(&mut self) -> Result<bool>;

    /// Get the raw field data of the current journal entry.
    ///
    /// Returns the raw string representing the key-value pairs of the journal entry, or `None` if there are no more entries.
    /// Returns an error if there was an issue reading the journal.
    fn get_entry_field_data(&mut self) -> Result<Option<JournalEntryRaw>>;

    /// Waits for the next journal entry to be available.
    ///
    /// This method should block until the next journal entry is available.
    fn wait_for_entry(&mut self) -> Result<()>;
}

/// Raw journal entry data.
///
/// This struct represents the raw data of a journal entry. It can be converted into a `JournalEntry` struct,
/// which contains the parsed key-value pairs of the journal entry.
#[derive(Debug)]
pub struct JournalEntryRaw {
    pub ts: DateTime<Utc>,
    pub fields: Vec<String>,
}

impl JournalEntryRaw {
    pub fn new(fields: Vec<String>, ts: DateTime<Utc>) -> Self {
        Self { ts, fields }
    }
}

/// An implementation of the `JournalRaw` trait that reads journal entries from the systemd journal.
///
/// This struct is used to read journal entries from the systemd journal. It relies on ffi calls into
/// libsystemd.
pub struct JournalRawImpl {
    journal: *mut sd_journal,
    wait_fd: PollFd,
    cursor_file: PathBuf,
}

impl JournalRawImpl {
    /// Timeout journal polling after 1 minute.
    const POLL_TIMEOUT_MS: i32 = 1000 * 60;
    const JOURNAL_CURSOR_FILE: &str = "JOURNALD_CURSOR";

    pub fn new(tmp_path: PathBuf) -> Self {
        let mut journal = std::ptr::null_mut();
        let cursor_file = tmp_path.join(Self::JOURNAL_CURSOR_FILE);
        let cursor_string = read_to_string(&cursor_file).ok();

        unsafe {
            sd_journal_open(&mut journal, 0);
        }

        let fd = unsafe { sd_journal_get_fd(journal) };
        let wait_fd = PollFd::new(fd, nix::poll::PollFlags::POLLIN);

        if let Some(cursor) = cursor_string {
            let cursor = cursor.trim();
            let ret = unsafe { sd_journal_seek_cursor(journal, cursor.as_ptr() as *const c_char) };
            if ret < 0 {
                warn!("Failed to seek journal to cursor: {}", ret);
            }
        } else if let Err(e) = Self::seek_to_current_boot_start(journal) {
            warn!("Couldn't seek journal to start of current boot: {}", e);
        }

        Self {
            journal,
            wait_fd,
            cursor_file,
        }
    }

    /// Seeks the journal to the start of the current boot's logs
    ///
    /// Returns OK if the journal is now set up to return the first log
    /// from the current boot
    /// Returns an error if the function could not confirm the next
    /// entry will be the start of the current boot
    fn seek_to_current_boot_start(journal: *mut sd_journal) -> Result<()> {
        let boot_id = read_system_boot_id()?;
        let boot_id_match = CString::new(format!("_BOOT_ID={}", boot_id.as_simple()))?;

        let ret =
            unsafe { sd_journal_add_match(journal, boot_id_match.as_ptr() as *const c_char, 0) };
        match ret {
            ret if ret < 0 => Err(eyre!(
                "Failed to add match on current boot ID to journal: {}",
                ret
            )),
            0 => Ok(()),
            _ => Ok(()),
        }
    }

    /// Save the current journal cursor to a file.
    ///
    /// This method saves the current journal cursor to a file so that the journal can be resumed from the
    /// same point after a restart.
    fn save_cursor(&self) {
        let mut cursor: MaybeUninit<*const c_char> = MaybeUninit::uninit();
        let ret = unsafe { sd_journal_get_cursor(self.journal, cursor.as_mut_ptr()) };
        if ret < 0 {
            warn!("Failed to get journal cursor: {}", ret);
        } else {
            let cursor = unsafe { cursor.assume_init() };
            let cursor_cstr = unsafe { std::ffi::CStr::from_ptr(cursor) };
            let cursor_str = cursor_cstr.to_str().unwrap_or_default();

            let write_result = std::fs::write(&self.cursor_file, cursor_str);
            unsafe {
                free(cursor as *mut libc::c_void);
            }
            match write_result {
                Ok(_) => (),
                Err(e) => warn!(
                    "Failed to write journal cursor to {:?}: {}",
                    self.cursor_file, e
                ),
            }
        }
    }

    pub fn get_timestamp(&self) -> Result<DateTime<Utc>> {
        let mut timestamp = 0u64;
        let ret = unsafe { sd_journal_get_realtime_usec(self.journal, &mut timestamp) };
        if ret < 0 {
            return Err(eyre!("Failed to get journal entry timestamp: {}", ret));
        }

        let datetime = NaiveDateTime::from_timestamp_micros(timestamp as i64)
            .ok_or_else(|| eyre!("Failed to convert journal timestamp to DateTime"))?;

        Ok(DateTime::<Utc>::from_utc(datetime, Utc))
    }
}

impl Drop for JournalRawImpl {
    fn drop(&mut self) {
        self.save_cursor();
        unsafe {
            libc::free(self.journal as *mut libc::c_void);
        }
    }
}

impl JournalRaw for JournalRawImpl {
    fn next_entry_available(&mut self) -> Result<bool> {
        let ret = unsafe { sd_journal_next(self.journal) };
        match ret {
            ret if ret < 0 => Err(eyre!("Failed to get next journal entry: {}", ret)),
            0 => Ok(false),
            _ => Ok(true),
        }
    }

    fn get_entry_field_data(&mut self) -> Result<Option<JournalEntryRaw>> {
        let mut data = MaybeUninit::uninit();
        let mut data_len = MaybeUninit::uninit();
        let mut fields = Vec::new();

        let timestamp = self.get_timestamp().map_or_else(
            |e| {
                debug!(
                    "Failed to get journal entry timestamp, falling back to ingestion time: {}",
                    e
                );
                Utc::now()
            },
            |t| t,
        );

        let mut enum_ret = unsafe {
            sd_journal_enumerate_data(self.journal, data.as_mut_ptr(), data_len.as_mut_ptr())
        };

        while enum_ret > 0 {
            let bytes =
                unsafe { std::slice::from_raw_parts(data.assume_init(), data_len.assume_init()) };

            let kv_string = String::from_utf8_lossy(bytes).to_string();
            fields.push(kv_string);

            enum_ret = unsafe {
                sd_journal_enumerate_data(self.journal, data.as_mut_ptr(), data_len.as_mut_ptr())
            };
        }

        if enum_ret < 0 {
            Err(eyre!("Failed to read journal entry data: {}", enum_ret))
        } else {
            Ok(Some(JournalEntryRaw::new(fields, timestamp)))
        }
    }

    fn wait_for_entry(&mut self) -> Result<()> {
        let mut fds = [self.wait_fd];
        let ret = poll(&mut fds, Self::POLL_TIMEOUT_MS)?;
        if ret < 0 {
            return Err(eyre!("Failed to poll for journal entry: {}", ret));
        }

        // This call clears the queue status of the poll fd
        let ret = unsafe { sd_journal_process(self.journal) };
        if ret < 0 {
            return Err(eyre!("Failed to process journal entry: {}", ret));
        }

        Ok(())
    }
}

/// A fully parsed journal entry.
///
/// This struct represents a fully parsed journal entry, with all key-value pairs parsed into a HashMap.
#[derive(Serialize, Debug)]
pub struct JournalEntry {
    pub ts: DateTime<Utc>,
    pub fields: HashMap<String, String>,
}

impl From<JournalEntryRaw> for JournalEntry {
    fn from(raw: JournalEntryRaw) -> Self {
        let fields = raw
            .fields
            .into_iter()
            .fold(HashMap::new(), |mut acc, field| {
                let kv: Vec<&str> = field.splitn(2, '=').collect();
                if kv.len() == 2 {
                    acc.insert(kv[0].to_string(), kv[1].to_string());
                }
                acc
            });

        Self { ts: raw.ts, fields }
    }
}

impl TryFrom<JournalEntry> for LogEntry {
    type Error = Error;

    fn try_from(mut entry: JournalEntry) -> Result<Self, Self::Error> {
        let ts = entry.ts;

        let fields = &mut entry.fields;
        let message = fields
            .remove("MESSAGE")
            .ok_or_else(|| eyre!("Journal entry is missing MESSAGE field"))?;

        let pid = fields.remove("_PID");
        let systemd_unit = fields.remove("_SYSTEMD_UNIT");
        let priority = fields.remove("PRIORITY");

        let extra_fields = entry
            .fields
            .into_iter()
            .fold(HashMap::new(), |mut acc, (k, v)| {
                acc.insert(k, LogValue::String(v));
                acc
            });
        let data = LogData {
            message,
            pid,
            systemd_unit,
            priority,
            original_priority: None,
            extra_fields,
        };

        Ok(LogEntry { ts, data })
    }
}

/// A wrapper around the `JournalRaw` trait that provides an iterator over journal entries.
///
/// This struct provides an iterator over journal entries, abstracting the raw journal reading logic.
pub struct Journal<J: JournalRaw> {
    journal: J,
}

impl<J: JournalRaw> Journal<J> {
    pub fn new(journal: J) -> Self {
        Self { journal }
    }

    /// Get the next journal entry.
    ///
    /// This function will return a `JournalEntry` if the next entry is available, or `None` if there are no more entries.
    /// Returns an error if there was an issue reading the journal.
    fn next_entry(&mut self) -> Result<Option<JournalEntry>> {
        match self.journal.next_entry_available()? {
            false => Ok(None),
            true => {
                let raw = self.journal.get_entry_field_data()?;
                Ok(raw.map(|raw| raw.into()))
            }
        }
    }

    /// Get an iterator over all available journal entries.
    pub fn iter(&mut self) -> impl Iterator<Item = JournalEntry> + '_ {
        std::iter::from_fn(move || match self.next_entry() {
            Ok(entry) => entry,
            Err(e) => {
                warn!("Failed to get next journal entry: {}", e);
                None
            }
        })
    }

    pub fn wait_for_entry(&mut self) -> Result<()> {
        self.journal.wait_for_entry()
    }
}

#[cfg(test)]
mod test {
    use super::*;

    use insta::{assert_json_snapshot, with_settings};
    use mockall::Sequence;

    #[test]
    fn test_from_raw_journal_entry() {
        let raw_entry = raw_journal_entry();
        let entry = JournalEntry::from(raw_entry);

        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(entry);
        });
    }

    #[test]
    fn test_journal_happy_path() {
        let mut raw_journal = MockJournalRaw::new();
        let mut seq = Sequence::new();

        raw_journal
            .expect_next_entry_available()
            .times(1)
            .in_sequence(&mut seq)
            .returning(|| Ok(true));

        raw_journal
            .expect_get_entry_field_data()
            .times(1)
            .in_sequence(&mut seq)
            .returning(|| Ok(Some(raw_journal_entry())));

        raw_journal
            .expect_next_entry_available()
            .times(1)
            .in_sequence(&mut seq)
            .returning(|| Ok(false));

        let mut journal = Journal::new(raw_journal);
        let mut journal_iter = journal.iter();
        let entry = journal_iter.next();
        assert!(entry.is_some());
        with_settings!({sort_maps => true}, {
            assert_json_snapshot!(entry.unwrap());
        });
        let entry = journal_iter.next();
        assert!(entry.is_none());
    }

    fn timestamp() -> DateTime<Utc> {
        let timestamp = NaiveDateTime::from_timestamp_millis(1713462571).unwrap();
        DateTime::<Utc>::from_utc(timestamp, Utc)
    }

    fn raw_journal_entry() -> JournalEntryRaw {
        let fields = [
            "_SYSTEMD_UNIT=user@1000.service",
            "MESSAGE=audit: type=1400 audit(1713462571.968:7508): apparmor=\"DENIED\" operation=\"open\" class=\"file\" profile=\"snap.firefox.firefox\" name=\"/etc/fstab\" pid=10122 comm=\"firefox\" requested_mask=\"r\" denied_mask=\"r\" fsuid=1000 ouid=0",
        ];

        JournalEntryRaw::new(fields.iter().map(|s| s.to_string()).collect(), timestamp())
    }
}
