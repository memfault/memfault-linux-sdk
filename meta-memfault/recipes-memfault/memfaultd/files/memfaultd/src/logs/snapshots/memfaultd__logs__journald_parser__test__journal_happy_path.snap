---
source: memfaultd/src/logs/journald_parser.rs
expression: entry.unwrap()
---
{
  "ts": "1970-01-20T19:57:42.571Z",
  "fields": {
    "MESSAGE": "audit: type=1400 audit(1713462571.968:7508): apparmor=\"DENIED\" operation=\"open\" class=\"file\" profile=\"snap.firefox.firefox\" name=\"/etc/fstab\" pid=10122 comm=\"firefox\" requested_mask=\"r\" denied_mask=\"r\" fsuid=1000 ouid=0",
    "_SYSTEMD_UNIT": "user@1000.service"
  }
}
